src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -292,92 +196,186 @@ internal sealed class MeterProviderSdk : MeterProvider
         }
         else
         {
-            this.listener.InstrumentPublished = (instrument, listener) =>
-            {
-                bool enabledMeasurements = false;
+            // Everything double
+            this.listener.SetMeasurementEventCallback<double>(this.MeasurementRecordedDoubleSingleStream);
+            this.listener.SetMeasurementEventCallback<float>((instrument, value, tags, state) => this.MeasurementRecordedDoubleSingleStream(instrument, value, tags, state));
+
+            // Everything long
+            this.listener.SetMeasurementEventCallback<long>(this.MeasurementRecordedLongSingleStream);
+            this.listener.SetMeasurementEventCallback<int>((instrument, value, tags, state) => this.MeasurementRecordedLongSingleStream(instrument, value, tags, state));
+            this.listener.SetMeasurementEventCallback<short>((instrument, value, tags, state) => this.MeasurementRecordedLongSingleStream(instrument, value, tags, state));
+            this.listener.SetMeasurementEventCallback<byte>((instrument, value, tags, state) => this.MeasurementRecordedLongSingleStream(instrument, value, tags, state));
+
+            this.listener.MeasurementsCompleted = (instrument, state) => this.MeasurementsCompletedSingleStream(instrument, state);
+        }
+
+        this.listener.Start();
+
+        OpenTelemetrySdkEventSource.Log.MeterProviderSdkEvent("MeterProvider built successfully.");
+    }
+
+    internal Resource Resource { get; }
+
+    internal List<object> Instrumentations => this.instrumentations;
+
+    internal MetricReader? Reader => this.reader;
+
+    internal int ViewCount => this.viewConfigs.Count;
+
+    internal object? InstrumentPublished(Instrument instrument, bool listeningIsManagedExternally)
+    {
+        var listenToInstrumentUsingSdkConfiguration = this.shouldListenTo(instrument);
+
+        if (listeningIsManagedExternally && listenToInstrumentUsingSdkConfiguration)
+        {
+            OpenTelemetrySdkEventSource.Log.MetricInstrumentIgnored(
+                instrument.Name,
+                instrument.Meter.Name,
+                "Instrument belongs to a Meter which has been enabled externally and via a subscription on the provider. External subscription will be ignored in favor of the provider subscription.",
+                "Don't call AddMeter when also using external management.");
```

Comment:
Probably not a huge deal, but this might cause a little confusion for meters added via `AddMeter` from our instrumentation. Like if someone is using .NET 8, uses appsettings.json to enable some metrics, and also calls `AddAspNetCoreInstrumentation()`. It may be unclear to peeps what caused this message.

Maybe since this is essentially a noop anyways, should we just not log anything in this case?

Reply:
I think we need to tighten up our story a bit here. Related to my comment https://github.com/open-telemetry/opentelemetry-dotnet/pull/4958#issuecomment-1806293792

Specifically:

> I think OTel's config can be complementary to the new config designed for IMetricsBuilder. I imagine it working like .NET's IConfiguration in that configuration can be sourced from multiple providers. OTel configuration would be the union of configuration sourced from environment variables, OTel config file, and appsettings.json.

I completely support config from multiple sources. As it stands though, configuring via `appsettings.json` and programmatically via `AddMeter` does not achieve the "union of configuration" I was envisioning.

I mentioned in this week's SIG meeting, I think our precedent to date has been that programatic configuration trumps other config source (e.g., environment variables). I'd like us to try and maintain a predictable precedent in light of IMetricsListener config.

My current thought is `AddMeter` should trump `appsettings.json`. This probably would mean that anything `AddMeter`ed would not be able to be enabled/disabled at runtime. I think this is ok. If folks want to use `appsettings.json` then they can migrate away from using `AddMeter`. I think this would begin to address this concern https://github.com/open-telemetry/opentelemetry-dotnet/pull/4958#issuecomment-1815383167.

I think we could introduce the ability for a user to explicitly declare the order of precedence for config sources at a later date.

Reply:
What you think @utpilla should we remove this? At the moment it logs a warning. Another idea: We could instead introduce a different log for this which does an info-level?

Reply:
Just had a zoom conversation with @CodeBlanch on this. It appears this _does_ achieve the "union of configuration". I just wasn't configuring things correctly. So all is good.

I also reevaluated my stance that we remove the log message. I think we keep it, but I shared some feedback to make it less confusing. Namely emphasize:

* That `AddMeter` and `AddInstrumentationXyz` can result in this log message
* That the SDKs programatic config always trumps other config sources by default

Reply:
I just pushed changes to the log message. There's a lot of information to convey so it was a bit of a challenge 😄

RE: The behavior above, to try and clarify for everyone...

When using this new `IMetricsBuilder` / `IMetricsListener` API there will be multiple `MeterListener`s active in the process. The SDK listener hasn't changed. If you disable some meter or instrument in the new API AND call `AddMeter` what happens is the `IMetricsListener` won't see the measurements. But the SDK listener will. In this way, actions performed on the SDK are always respected. The only special case that was added here is if BOTH `IMetricsListener` is subscribed AND the SDK is subscribed then we ignore the "external" subscription (to prevent duplicate measurements).